zerver/migrations/0478_add_topic_wildcard_mentioned_flag_and_create_index.py

```diff
@@ -0,0 +1,119 @@
+    for user_message in user_messages:
+        if user_message.flags.topic_wildcard_mentioned or user_message.flags.group_mentioned:
+            user_message.flags.topic_wildcard_mentioned = False
+            user_message.flags.group_mentioned = False
+            user_messages_to_update.add(user_message)
+
+    UserMessage.objects.bulk_update(user_messages_to_update, ["flags"])
+
+
+class Migration(migrations.Migration):
+    atomic = False
+
+    dependencies = [
+        ("zerver", "0477_alter_realmuserdefault_automatically_follow_topics_policy_and_more"),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name="archivedusermessage",
+            name="flags",
+            field=bitfield.models.BitField(
+                [
+                    "read",
+                    "starred",
+                    "collapsed",
+                    "mentioned",
+                    "stream_wildcard_mentioned",
+                    "topic_wildcard_mentioned",
+                    "group_mentioned",
+                    "force_expand",
+                    "force_collapse",
+                    "has_alert_word",
+                    "historical",
+                    "is_private",
+                    "active_mobile_push_notification",
+                ],
+                default=0,
+            ),
+        ),
+        migrations.AlterField(
+            model_name="usermessage",
+            name="flags",
+            field=bitfield.models.BitField(
+                [
+                    "read",
+                    "starred",
+                    "collapsed",
+                    "mentioned",
+                    "stream_wildcard_mentioned",
+                    "topic_wildcard_mentioned",
+                    "group_mentioned",
+                    "force_expand",
+                    "force_collapse",
+                    "has_alert_word",
+                    "historical",
+                    "is_private",
+                    "active_mobile_push_notification",
+                ],
+                default=0,
+            ),
+        ),
+        AddIndexConcurrently(
+            model_name="usermessage",
+            index=models.Index(
+                "user_profile",
+                "message",
+                condition=models.Q(
+                    ("flags__andnz", 8),
+                    ("flags__andnz", 16),
+                    ("flags__andnz", 32),
+                    ("flags__andnz", 64),
+                    _connector="OR",
+                ),
+                name="zerver_usermessage_any_mentioned_message_id",
+            ),
+        ),
+        RemoveIndexConcurrently(
+            model_name="usermessage",
+            name="zerver_usermessage_wildcard_mentioned_message_id",
+        ),
+        migrations.RunPython(
+            clear_set_bit_topic_wildcard_and_group_mentioned,
+            reverse_code=migrations.RunPython.noop,
+            elidable=True,
+        ),
```

Comment:
This sort of complex migration is generally best done in Zulip Cloud in multiple staged deployments, and for that reason we should split up this commit and its migrations. The sequence I would recommend is:
* First, the `AlterField` entries (very cheap) and `AddIndexConcurrently`
* Then, the backfill function `clear_set_bit_topic_wildcard_and_group_mentioned`.
* And finally, I think we should skip removing `zerver_usermessage_wildcard_mentioned_message_id` at all in this PR -- we will want to do it as a follow-up once we have confirmed that we're not going to have to revert this migration for some reason.

Reply:
That looks like a reasonable ordering to me.  Splitting out the commit to do the backfill would also be useful to better highlight the _why_ of it -- specifically, in the first commit we can call out in the AlterField commit that we're renaming two now-unused and historical bits of the bitfield, but they may have old data that we need to clear.  The third commit can be explicit that we're clearing out old data from the now-renamed bitfields, which is why we need to walk the entire table.